### PR TITLE
fix(jax): add hessian energy loss

### PR DESCRIPTION
**Pull request by njzjz**
## Summary
- add Hessian prefactors directly to dpmodel `EnergyLoss` instead of introducing a separate loss class
- add Hessian label requirements, loss/RMSE reporting, and serialization fields
- enable Hessian outputs in the JAX trainer when the Hessian loss is configured

## Tests
- `source venv/bin/activate && pytest source/tests/common/dpmodel/test_loss_ener.py -q`
- `source venv/bin/activate && ruff check .`
- `source venv/bin/activate && ruff format .`

## Summary by CodeRabbit

* **New Features**
  * Optional energy Hessian-loss supervision with configurable scheduled weighting and Hessian RMSE/MAE reporting.
  * Trainer and model now automatically activate Hessian outputs based on the loss/data contract, including Hessian tensor creation for mixed/padded batches.
* **Bug Fixes**
  * Hessian activation is idempotent to prevent repeated reconfiguration.
  * Clearer runtime validation for unsupported Hessian + huber combinations.
* **Tests**
  * Expanded Hessian loss, serialization (v4→v5 backward compatible), padding/ghost handling, and MAE/L2 dispatch coverage across NumPy and PyTorch.
* **Documentation**
  * Updated Hessian training docs for JAX and clarified `ener`/legacy `ener_hess` alias behavior.

### CODE REVIEW

File: deepmd/jax/train/trainer.py [hunk 1]
```diff
@@ -117,6 +117,8 @@ def get_lr_and_coef(lr_param: dict) -> LearningRateExp:
         loss_type = loss_param.get("type", "ener")
         if loss_type == "ener":
             self.loss = EnergyLoss.get_loss(loss_param)
+            if getattr(self.loss, "has_h", False):
+                self.model.enable_hessian()
```
[the remark this answers is not in the document]
Comment (njzjz) [reply deepmd/jax/train/trainer.py]: Addressed in commits 1550599 to ae453b0. Sorry, I missed the inline reply when resolving this thread.

**Comment by njzjz** [comment]
Possible reviewers based on changed lines, exact file history, and exact-file review history:

- @wanghan-iapcm — 15 commits on changed files; 26 reviews on exact changed files (`deepmd/dpmodel/loss/ener.py`, `deepmd/jax/train/trainer.py`, `deepmd/utils/data.py`, `deepmd/utils/data_system.py`, `source/tests/common/dpmodel/test_loss_ener.py`).
- @OutisLi — 18 reviews on exact changed files (`deepmd/dpmodel/loss/ener.py`, `deepmd/jax/train/trainer.py`).

No review request was made automatically.

Coding agent: Codex
Codex version: codex-cli 0.144.4
Model: gpt-5.6-sol
Reasoning effort: xhigh

File: deepmd/pt/loss/ener.py [hunk 2]
```diff
@@ -775,6 +786,37 @@ def forward(
...
+            find_hessian = label.get("find_hessian", 0.0)
+            pref_h = pref_h * find_hessian
+            if maskf is not None:
+                hessian_shape = (_nf, _nloc * 3, _nloc * 3)
+                diff_h = label["hessian"].reshape(hessian_shape) - model_pred[
+                    "hessian"
+                ].reshape(hessian_shape)
+                # Both Cartesian axes must refer to real atoms for a Hessian
+                # element to contribute to the loss or display denominator.
+                l2_hessian_loss = masked_pair_mean(torch.square(diff_h), maskf, ncomp=3)
```
Comment (wanghan-iapcm) [inline deepmd/pt/loss/ener.py:804]: This is CodeRabbit's open thread from 24 Jul, and I agree it's real - but it needs fixing in both backends, not just here.

Every other supervised term in this class dispatches on `self.loss_func` (`mse` / `mae` / `else: raise NotImplementedError`). The Hessian term hard-codes squared error, so `loss_func: "mae"` trains energy, force and virial on L1 while the Hessian minimises L2 inside the same scalar. With Hessian residuals of order 10 that term can dominate the objective, silently, in the opposite direction to what was configured. `mae_h` is computed only for display, so nothing surfaces it.

Same gap for Huber: the `use_huber` guard in `__init__` rejects `has_pf`/`has_gf`/`relative_f` but not `has_h`, so a Huber-configured run gets a raw MSE Hessian term with no error.

The identical block exists at `deepmd/dpmodel/loss/ener.py:765` and has the same defect - it wasn't flagged, but fixing only the PyTorch side would leave the two backends disagreeing on what `loss_func` means. Please add the `mse`/`mae`/`NotImplementedError` branch in both, and add `has_h` to the Huber incompatibility check.

File: deepmd/dpmodel/loss/ener.py [hunk 3]
```diff
@@ -819,6 +857,17 @@ def label_requirement(self) -> list[DataRequirementItem]:
...
             )
+        if self.has_h:
+            label_requirement.append(
+                DataRequirementItem(
+                    "hessian",
+                    ndof=1,
+                    atomic=False,
+                    must=False,
+                    high_prec=False,
+                    special_shape="hessian",
```
Comment (wanghan-iapcm) [inline deepmd/dpmodel/loss/ener.py:879]: Declaring the Hessian `atomic=False` sends it down the wrong branch of `_merge_batch_data` for mixed-system batches. That path does `np.concatenate([bb[kk] for bb in batch_data], axis=0)`, and each frame's Hessian is `(1, 9*N_i**2)` with a different `N_i` per system, so a `batch_size: "mixed:N"` run raises `ValueError: all the input array dimensions except for the concatenation axis must match exactly`.

Worth noting where this bites: PyTorch never reaches `_merge_batch_data` (its dataloader builds per-system loaders and doesn't parse `mixed:N`), but the JAX trainer calls `DeepmdDataSystem.get_batch()` directly - so this is reachable precisely in the JAX + mixed-batching + Hessian combination this PR is adding.

Switching to `atomic=True` wouldn't fix it either: that branch copies a flat prefix, whereas a correct pad has to embed the `3N_i x 3N_i` block in the top-left of a `3*max_natoms` square - which is what your own `_padded_hessian` test helper does. Since `special_shape="hessian"` is the shape contract this PR introduces, `_merge_batch_data` seems like the right place to honour it (or to raise a clear `NotImplementedError` for now).

Related coverage gap: the new padding tests inject pre-padded `label["hessian"]` tensors directly, so the real merge path is never exercised - which is why this doesn't show up as a failure.

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]
Folding Hessian supervision into the shared `EnergyLoss` is the right call, and inverting the control flow so the loss's `label_requirement` drives `enable_hessian()` is a much better contract than the trainer sniffing `start_pref_h`. I checked `masked_pair_mean` numerically rather than by eye: the pair mask really is the outer product of the per-atom mask, the denominator is `(3*n_real)**2`, an all-padding frame returns a neutral 0.0, and it reduces exactly to `mean` under an all-ones mask. The `enable_hessian()` idempotency guards are correct on both backends, and with both prefs left at 0 the whole thing is inert.

Two things I'd like resolved before this goes in, both inline.

Non-blocking notes:

1. `@version` stays at 4 in both `serialize()` implementations while two new fields are added, and `check_version_compatibility(version, 4, 1)` is unchanged. Because both `__init__`s end in `**kwargs`, an older deepmd-kit reading a Hessian-enabled dict passes the version gate and then silently discards `start_pref_h`/`limit_pref_h` - the model loads with no Hessian term and no error. Worth a bump plus an explicit 4-to-5 branch.

2. `has_h` picks up `or inference`, which is consistent with how `has_e`/`has_f`/`has_v`/`has_ae`/`has_pf` already behave - but every one of those labels is linear in N, whereas the Hessian is `9N**2`. `dp change-bias` builds the loss with `inference=True` purely to harvest `label_requirement`, so it now requests a Hessian it never uses, and `_load_single_data` zero-fills it at roughly 72 MB per frame for a 1000-atom system (vs 24 KB for force). The convention is fine; it just doesn't extend safely to a quadratic label. Excluding the Hessian item when `has_h` is true only because of `inference` would fix it.

3. `deepmd/pd` is untouched, so after this lands the same input.json trains a Hessian term on pt/jax/dpmodel but not on pd, which keeps the old AND semantics. Related: `alias=["ener_hess"]` makes `type: ener_hess` a valid input for the first time, and it now normalizes to `ener` and reaches pd's un-audited path with no pd test.

4. `doc/model/train-energy-hessian.md` still advertises PyTorch only, though this PR wires and tests JAX end-to-end.

5. The `and` -> `or` change means an existing `start_pref_h: 100, limit_pref_h: 0` config now trains with an annealed Hessian term where it previously trained with none. I think the new behaviour is right - argcheck's shared `start_pref()` text has always said the term is ignored only when *both* are zero, so the old AND was the outlier - but it changes results for existing configs silently and deserves a release note.

File: deepmd/pt/loss/ener.py (hunk 2 above)
Comment (njzjz) [reply deepmd/pt/loss/ener.py:804]: Fixed in both backends in `a663f9242` (branch also rebased onto master).

`deepmd/pt/loss/ener.py` and `deepmd/dpmodel/loss/ener.py` now dispatch the Hessian term the same way every other supervised term does: `mse` adds `pref_h * l2_hessian_loss`, `mae` adds `pref_h * mae_hessian_loss`, anything else raises `NotImplementedError("... is not implemented for hessian loss")`. The MAE reduction reuses the same `masked_pair_mean(..., ncomp=3)` / plain-mean split as the MSE one, so the placeholder-pair masking is unchanged; `rmse_h` stays MSE-derived for display, and `mae_h` is now computed once and shared between the `mae` display flag and the training term.

Also added `has_h` to the Huber incompatibility check in both `__init__`s, next to the existing `has_pf`/`has_gf`/`relative_f` guard.

Regressions in `TestDPModelEnergyLossHessianGradAccum`: a constant-residual frame where L1 and L2 are distinguishable (residual 10 → loss 10 under `mae`, 100 under `mse`), asserted for dpmodel and pt; an unknown `loss_func` raising through the Hessian branch; and both constructors rejecting `use_huber` with `start_pref_h > 0`.

File: deepmd/dpmodel/loss/ener.py (hunk 3 above)
Comment (njzjz) [reply deepmd/dpmodel/loss/ener.py:879]: Fixed in `a663f9242` by honouring the `special_shape="hessian"` contract in `_merge_batch_data`, as you suggested.

`_merge_batch_data` now has a third branch ahead of the atomic/non-atomic split: for `special_shape == "hessian"` it allocates `(nbatch, 3*max_natoms, 3*max_natoms)` and writes each frame's `3*N_i x 3*N_i` block into the top-left corner, leaving the padded rows and columns zero. That matches what `_padded_hessian` does in the test helper, and the padded entries are exactly the ones `masked_pair_mean` drops, so the loss is unaffected.

New `source/tests/common/test_data_system_hessian.py` builds two on-disk systems with 2 and 3 atoms and a distinct constant Hessian each. It goes through the real loader — no pre-padded tensors — and calls `_merge_batch_data` with one frame from each system, so the ragged case is hit deterministically (`get_batch_mixed` picks systems at random). On the unpatched code it reproduces your `ValueError: ... the array at index 0 has size 36 and the array at index 1 has size 81`; after the fix it asserts each block lands in its corner and everything outside is zero. A second test loops `get_batch()` 20 times so the public entry point is covered for whichever combination is sampled.

File: deepmd/pt/loss/ener.py [hunk 4]
```diff
@@ -775,6 +786,37 @@ def forward(
...
+            more_loss["rmse_h"] = self.display_if_exist(
+                l2_hessian_loss.sqrt().detach(), find_hessian
+            )
+            if mae:
+                if maskf is not None:
+                    mae_h = masked_pair_mean(torch.abs(diff_h), maskf, ncomp=3)
+                else:
+                    mae_h = torch.mean(torch.abs(diff_h))
+                more_loss["mae_h"] = self.display_if_exist(mae_h.detach(), find_hessian)
+
```
[the remark this answers is not in the document]
Comment (njzjz) [reply deepmd/pt/loss/ener.py:833]: Confirmed and fixed in `a663f9242`, in both backends — `deepmd/dpmodel/loss/ener.py` had the identical defect. The Hessian term now branches on `self.loss_func` (`mse` / `mae` / `NotImplementedError`) like every other supervised term, and `use_huber` together with `has_h` is rejected in `__init__` rather than silently producing a raw MSE Hessian term.

**Review by wanghan-iapcm — APPROVED** [review]
Both blocking items are fixed, and I verified them at `a1ed51d2a` rather than from the replies.

The Hessian term now dispatches on `loss_func` in both backends ([dpmodel](https://github.com/deepmodeling/deepmd-kit/blob/a1ed51d2a55427223e6c4c44ec791d60ce08db94/deepmd/dpmodel/loss/ener.py#L780-L787), [pt](https://github.com/deepmodeling/deepmd-kit/blob/a1ed51d2a55427223e6c4c44ec791d60ce08db94/deepmd/pt/loss/ener.py#L817-L824)), with `has_h` added to the Huber guard in both constructors. The `_merge_batch_data` branch is in the right place -- ahead of the atomic/non-atomic split, embedding each frame's `3N_i x 3N_i` block in the top-left of the padded square ([data_system.py](https://github.com/deepmodeling/deepmd-kit/blob/a1ed51d2a55427223e6c4c44ec791d60ce08db94/deepmd/utils/data_system.py#L557-L572)).

The regression tests genuinely fail pre-fix, which is what I wanted to see. `test_mae_loss_func_uses_l1_for_the_hessian` zeroes every prefactor except `pref_h` and uses a constant residual of 10, so the old code contributes 100 where it asserts 10, and it cross-checks that the mse config still gives 100. `test_data_system_hessian.py` goes through the real loader with 2- and 3-atom systems, so the old `np.concatenate` hits `(1, 36)` against `(1, 81)`. The unknown-`loss_func` and `use_huber` cases are covered in both backends. Good coverage of the exact intersections that were missing.

Two of my earlier non-blocking notes are still open. Neither blocks this, but flagging that they were not answered either way:

1. `@version` is still 4 in both `serialize()` implementations ([dpmodel](https://github.com/deepmodeling/deepmd-kit/blob/a1ed51d2a55427223e6c4c44ec791d60ce08db94/deepmd/dpmodel/loss/ener.py#L893-L895), [pt](https://github.com/deepmodeling/deepmd-kit/blob/a1ed51d2a55427223e6c4c44ec791d60ce08db94/deepmd/pt/loss/ener.py#L934-L936)) while `start_pref_h`/`limit_pref_h` are now emitted conditionally, and `check_version_compatibility(version, 4, 1)` is unchanged. Since both `__init__`s end in `**kwargs`, an older deepmd-kit reading a Hessian-enabled dict passes the version gate and then silently drops both fields -- the model loads with no Hessian term and no error. This is the case the repo's serialization-version discipline is meant to catch, so a bump plus an explicit 4-to-5 branch would be worth doing before or shortly after this lands.

2. [`pt/loss/ener.py#L166`](https://github.com/deepmodeling/deepmd-kit/blob/a1ed51d2a55427223e6c4c44ec791d60ce08db94/deepmd/pt/loss/ener.py#L164-L168) still reads `self.has_h = (start_pref_h != 0.0 or limit_pref_h != 0.0) or inference`. `dp change-bias` constructs the loss with `inference=True` only to harvest `label_requirement`, so it now requests a Hessian it never uses and zero-fills roughly 9N^2 per frame -- about 72 MB per frame at 1000 atoms, against 24 KB for force. The convention matches the other `has_*` flags; it just does not extend safely to a quadratic label.

Thanks for the doc cleanup to the `loss[ener]` references as well.

**Comment by njzjz** [comment]
@wanghan-iapcm Addressed the remaining non-blocking notes in 4d667b290 and resolved the latest master conflict in eb42bec72:

- Hessian-bearing energy-loss payloads now use serialization version 5, while ordinary energy losses remain version 4 for compatibility. Loading versions 1–4 explicitly defaults the Hessian prefactors to zero.
- An inference-only PyTorch loss no longer requests the quadratic Hessian label, so dp change-bias avoids allocating an unused 9N² zero array.
- Paddle now uses the same OR activation semantics for Hessian schedules, accepts the legacy ener_hess alias, and declares the full Hessian shape contract.
- The Hessian documentation now covers JAX training, fine-tuning, testing, freezing, and the one-zero-prefactor behavior change.

Validation after merging current master: 20 focused loss, serialization, argument-schema, and JAX tests passed; the complete PyTorch change-bias scenario passed; ruff format . and ruff check . passed. A Paddle regression test is included, but Paddle is not installed in the local environment.

Coding agent: Codex
Codex version: codex-cli 0.144.6
Model: gpt-5.6-sol
Reasoning effort: xhigh

File: deepmd/pd/loss/ener.py [hunk 5]
```diff
@@ -649,7 +649,9 @@ def __init__(
             Other keyword arguments.
         """
         super().__init__(**kwargs)
-        self.has_h = (start_pref_h != 0.0 and limit_pref_h != 0.0) or self.inference
+        # A scheduled term is active when either endpoint is nonzero. Requiring
+        # both endpoints silently disabled valid ramp-up and ramp-down inputs.
+        self.has_h = (start_pref_h != 0.0 or limit_pref_h != 0.0) or self.inference
```
[the remark this answers is not in the document]
Comment (njzjz) [reply deepmd/pd/loss/ener.py:681]: Fixed in 074e277fb. `EnergyHessianStdLoss` now rejects `use_huber=True` after computing `has_h`, using the same runtime error as the dpmodel and PyTorch implementations. A Paddle regression test covers the unsupported combination.

Validation: `ruff format .`, `ruff check .`, Python syntax compilation, and diff checks passed. Paddle is not installed in the local environment, so execution of the Paddle-specific test remains CI-validated.

Coding agent: Codex
Codex version: codex-cli 0.144.6
Model: gpt-5.6-sol
Reasoning effort: xhigh

File: deepmd/pd/loss/ener.py [hunk 6]
```diff
@@ -701,10 +703,11 @@ def label_requirement(self) -> list[DataRequirementItem]:
             label_requirement.append(
                 DataRequirementItem(
                     "hessian",
-                    ndof=1,  # 9=3*3 --> 3N*3N=ndof*natoms*natoms
-                    atomic=True,
+                    ndof=1,
+                    atomic=False,
                     must=False,
                     high_prec=False,
+                    special_shape="hessian",
```
[the remark this answers is not in the document]
Comment (njzjz) [reply deepmd/pd/loss/ener.py:751]: Fixed in 074e277fb. Paddle now expands the atom mask over Cartesian components, applies it on both Hessian axes, and normalizes each frame by the square of its real Cartesian degrees of freedom. The same mask is used for the optional Hessian MAE metric. The regression test places large residuals in padded blocks and verifies that only the real-real 3x3 block contributes.

Validation: `ruff format .`, `ruff check .`, Python syntax compilation, and diff checks passed. Paddle is not installed in the local environment, so execution of the Paddle-specific test remains CI-validated.

Coding agent: Codex
Codex version: codex-cli 0.144.6
Model: gpt-5.6-sol
Reasoning effort: xhigh
